File: python/graphs/dfs_production.py
from collections.abc import Iterable

Cell = tuple[int, int]


def dfs_grid_path(
    width: int,
    height: int,
    walls: Iterable[Cell],
    start: Cell,
    target: Cell,
) -> list[Cell]:
    if width <= 0 or height <= 0:
        raise ValueError("grid dimensions must be positive")

    def inside(cell: Cell) -> bool:
        row, col = cell
        return 0 <= row < height and 0 <= col < width

    if not inside(start) or not inside(target):
        raise ValueError("start and target must be inside the grid")

    total = width * height
    start_id = start[0] * width + start[1]
    target_id = target[0] * width + target[1]
    blocked = bytearray(total)

    for row, col in walls:
        if inside((row, col)):
            blocked[row * width + col] = 1

    if blocked[start_id] or blocked[target_id]:
        return []

    visited = bytearray(total)
    parent = [-1] * total
    stack = [start_id]
    visited[start_id] = 1

    while stack:
        cell_id = stack.pop()

        if cell_id == target_id:
            path: list[Cell] = []
            while cell_id != -1:
                path.append(divmod(cell_id, width))
                cell_id = parent[cell_id]
            path.reverse()
            return path

        row, col = divmod(cell_id, width)
        for next_row, next_col in ((row - 1, col), (row, col + 1), (row + 1, col), (row, col - 1)):
            if 0 <= next_row < height and 0 <= next_col < width:
                next_id = next_row * width + next_col
                if not blocked[next_id] and not visited[next_id]:
                    visited[next_id] = 1
                    parent[next_id] = cell_id
                    stack.append(next_id)

    return []

File: java/graphs/DfsProduction.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.List;

public class DfsProduction {
    record Cell(int row, int col) {}

    public static List<Cell> dfsGridPath(boolean[][] walls, Cell start, Cell target) {
        int height = walls.length;
        int width = walls[0].length;

        if (!inside(start, width, height) || !inside(target, width, height)) {
            throw new IllegalArgumentException("start and target must be inside the grid");
        }

        if (walls[start.row()][start.col()] || walls[target.row()][target.col()]) {
            return List.of();
        }

        int total = width * height;
        boolean[] visited = new boolean[total];
        int[] parent = new int[total];
        java.util.Arrays.fill(parent, -1);
        Deque<Integer> stack = new ArrayDeque<>();

        int startId = start.row() * width + start.col();
        int targetId = target.row() * width + target.col();
        stack.push(startId);
        visited[startId] = true;

        while (!stack.isEmpty()) {
            int cellId = stack.pop();
            Cell cell = new Cell(cellId / width, cellId % width);

            if (cellId == targetId) {
                List<Cell> path = new ArrayList<>();
                for (int currentId = targetId; currentId != -1; currentId = parent[currentId]) {
                    path.add(new Cell(currentId / width, currentId % width));
                }
                Collections.reverse(path);
                return path;
            }

            int[][] moves = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
            for (int[] move : moves) {
                int nextRow = cell.row() + move[0];
                int nextCol = cell.col() + move[1];
                if (nextRow >= 0 && nextRow < height && nextCol >= 0 && nextCol < width && !walls[nextRow][nextCol]) {
                    int nextId = nextRow * width + nextCol;
                    if (!visited[nextId]) {
                        visited[nextId] = true;
                        parent[nextId] = cellId;
                        stack.push(nextId);
                    }
                }
            }
        }

        return List.of();
    }

    private static boolean inside(Cell cell, int width, int height) {
        return cell.row() >= 0 && cell.row() < height && cell.col() >= 0 && cell.col() < width;
    }
}
